### PR TITLE
Adding configurabele TCP buffers and minor hafixes

### PR DESCRIPTION
**Pull request by Markuze**
- [Network] set socket RX/TX Buffers

Adding configurable TCP buffers for inbound and outbound TCP connections.
This change will improve networking performance, especially for long RTT connections.

### Test Plan

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4649)

### CODE REVIEW

File: config/src/config/network_config.rs [hunk 1]
```diff
@@ -51,6 +51,10 @@ pub const MAX_MESSAGE_SIZE: usize = 64 * 1024 * 1024; /* 64 MiB */
 pub const CONNECTION_BACKOFF_BASE: u64 = 2;
 pub const IP_BYTE_BUCKET_RATE: usize = 102400 /* 100 KiB */;
 pub const IP_BYTE_BUCKET_SIZE: usize = IP_BYTE_BUCKET_RATE;
+pub const INBOUND_TCP_RX_BUFFER_SIZE: u32 = 512 * 1024; // 3MB ~6MB/s with 500ms latency
+pub const INBOUND_TCP_TX_BUFFER_SIZE: u32 = 512 * 1024; // 1MB use a bigger spoon
```
Comment (igor-aptos) [inline config/src/config/network_config.rs:55]: these two comments don't seem to match constant values, if I understand things correctly?

**Review by igor-aptos — APPROVED** [review]

File: config/src/config/network_config.rs [hunk 2]
```diff
@@ -79,6 +83,10 @@ pub struct NetworkConfig {
     pub mutual_authentication: bool,
     pub network_id: NetworkId,
     pub runtime_threads: Option<usize>,
+    pub inbound_rx_buffer_size_bytes: Option<u32>,
+    pub inbound_tx_buffer_size_bytes: Option<u32>,
+    pub outbound_rx_buffer_size_bytes: Option<u32>,
+    pub outbound_tx_buffer_size_bytes: Option<u32>,
```
Comment (msmouse) [inline config/src/config/network_config.rs:89]: I heard `Option`s are not friendly to terraform templating, etc, shall we just default to the default numbers?

Comment (msmouse) [reply config/src/config/network_config.rs:89]: @rustielin  @sherry-x ?

File: developer-docs-site/docs/nodes/validator-node/operator/node-requirements.md [hunk 3]
```diff
@@ -48,7 +48,11 @@ The current hardware requirements are set considering the estimated growth over
 
 **Local SSD vs. network storage**
 
-Cloud deployments require choosing between using local or network storage such as AWS EBS, GCP PD. Local SSD provides lower latency and cost, especially relative to IOPS. 
+<<<<<<< HEAD
+Cloud deployments require choosing between using local or network storage such as AWS EBS, GCP PD. Local SSD provides lower latency and cost, especially relative to IOPS.
+=======
```
Comment (perryjrandall) [inline developer-docs-site/docs/nodes/validator-node/operator/node-requirements.md:53]: merge conflict

File: testsuite/forge.py [hunk 4]
```diff
@@ -1598,6 +1599,7 @@ def test(
             outputs.append(ForgeFormatter(github_step_summary, format_comment))
         forge_context.report(result, outputs)
 
+        print(forge_args)
```
Comment (perryjrandall) [inline testsuite/forge.py:1602]: remove this

File: testsuite/forge.py [hunk 5]
```diff
@@ -1215,6 +1215,7 @@ def create_forge_command(
         forge_args.append("--keep")
     if forge_enable_haproxy == "true":
         forge_args.append("--enable-haproxy")
+    forge_args.append("--enable-haproxy")
```
Comment (perryjrandall) [inline testsuite/forge.py:1218]: remove this

File: testsuite/forge.py [hunk 6]
```diff
@@ -1185,9 +1185,9 @@ def create_forge_command(
         forge_args.extend([
             "--duration-secs", forge_runner_duration_secs
         ])
-
     if forge_num_validators:
         forge_args.extend(["--num-validators", forge_num_validators])
+    forge_args.extend(["--num-validators", "100"])
```
Comment (perryjrandall) [inline testsuite/forge.py:1190]: remove

File: developer-docs-site/docs/nodes/validator-node/operator/node-requirements.md [hunk 7]
```diff
@@ -48,7 +48,11 @@ The current hardware requirements are set considering the estimated growth over
 
 **Local SSD vs. network storage**
 
-Cloud deployments require choosing between using local or network storage such as AWS EBS, GCP PD. Local SSD provides lower latency and cost, especially relative to IOPS. 
+<<<<<<< HEAD
```
Comment (JoshLind) [inline developer-docs-site/docs/nodes/validator-node/operator/node-requirements.md]: Can you fix the merge conflict?

File: network/builder/src/builder.rs [hunk 8]
```diff
@@ -90,9 +92,11 @@ impl NetworkBuilder {
         inbound_connection_limit: usize,
         inbound_rate_limit_config: Option<RateLimitConfig>,
         outbound_rate_limit_config: Option<RateLimitConfig>,
+        tcp_buffer_cfg: TCPBufferCfg,
     ) -> Self {
         // A network cannot exist without a PeerManager
-        // TODO:  construct this in create and pass it to new() as a parameter. The complication is manual construction of NetworkBuilder in various tests.
+        // TODO:  construct this in create and pass it to new() as a parameter. The complication is manual construction of NetworkBuilder in various tests. +1
+        // Change here
```
Comment (JoshLind) [inline network/builder/src/builder.rs:99]: ?

File: network/src/peer_manager/builder.rs [hunk 9]
```diff
@@ -278,11 +283,16 @@ impl PeerManagerBuilder {
             ),
         };
 
+        //fucking here
```
Comment (JoshLind) [inline network/src/peer_manager/builder.rs:286]: lol? Please remove 😄

File: network/src/peer_manager/builder.rs [hunk 10]
```diff
@@ -406,6 +416,14 @@ impl PeerManagerBuilder {
             .add_connection_event_listener()
     }
 
+    pub fn get_tcp_buffers_cfg(&self) -> TCPBufferCfg {
+        self.peer_manager_context
+            .as_ref()
+            .expect("Cannot add an event listener if PeerManager has already been built.")
+            .tcp_buffer_cfg
+        //       .clone()
```
Comment (JoshLind) [inline network/src/peer_manager/builder.rs:424]: ?

File: testsuite/forge.py (hunk 6 above)
Comment (JoshLind) [inline testsuite/forge.py]: I don't think we want to hard code this, right?

File: testsuite/forge.py (hunk 5 above)
Comment (JoshLind) [inline testsuite/forge.py]: Likewise here.

File: network/src/transport/mod.rs [hunk 11]
```diff
@@ -54,6 +54,8 @@ pub const APTOS_TCP_TRANSPORT: tcp::TcpTransport = tcp::TcpTransport {
     ttl: None,
     // Use TCP_NODELAY for Aptos tcp connections.
     nodelay: Some(true),
+    // Use default TCP setting, overriden by Network config
```
Comment (JoshLind) [inline network/src/transport/mod.rs:57]: nit: overriden -> overridden

**Review by JoshLind — COMMENTED** [review]
Thanks @Markuze! Looks like the PR still needs to be cleaned up a bit 😄

File: config/src/config/network_config.rs (hunk 2 above)
Comment (rustielin) [reply config/src/config/network_config.rs:89]: No it should be fine, so long as this lands: https://github.com/aptos-labs/aptos-core/pull/4847

File: network/netcore/src/transport/tcp.rs [hunk 12]
```diff
@@ -146,13 +191,26 @@ async fn resolve_with_filter(
...
+    tcp_buff_cfg: TCPBufferCfg,
+) -> io::Result<TcpStream> {
     let protos = addr.as_slice();
 
     if let Some(((ipaddr, port), _addr_suffix)) = parse_ip_tcp(protos) {
         // this is an /ip4 or /ip6 address, so we can just connect without any
         // extra resolving or filtering.
-        TcpStream::connect((ipaddr, port)).await
+        let addr = SocketAddr::new(ipaddr, port);
+        let socket = tokio::net::TcpSocket::new_v4()?;
```
Comment (zekun000) [inline network/netcore/src/transport/tcp.rs:204]: what about v6 addr?

**Review by zekun000 — COMMENTED** [review]
can we update the comments and address all nits?

File: network/src/peer_manager/builder.rs (hunk 10 above)
Comment (zekun000) [reply network/src/peer_manager/builder.rs:424]: and this

File: config/src/config/network_config.rs (hunk 1 above)
Comment (zekun000) [reply config/src/config/network_config.rs:55]: need to update comments

**Review by zekun000 — APPROVED** [review]

File: config/src/config/network_config.rs [hunk 13]
```diff
@@ -51,6 +51,10 @@ pub const MAX_MESSAGE_SIZE: usize = 64 * 1024 * 1024; /* 64 MiB */
 pub const CONNECTION_BACKOFF_BASE: u64 = 2;
 pub const IP_BYTE_BUCKET_RATE: usize = 102400 /* 100 KiB */;
 pub const IP_BYTE_BUCKET_SIZE: usize = IP_BYTE_BUCKET_RATE;
+pub const INBOUND_TCP_RX_BUFFER_SIZE: u32 = 3 * 1024 * 1024; // 3MB ~6MB/s with 500ms latency
+pub const INBOUND_TCP_TX_BUFFER_SIZE: u32 = 512 * 1024; // 1MB use a bigger spoon
```
[the remark this answers is not in the document]
Comment (Markuze) [reply config/src/config/network_config.rs:55]: Informal but descriptive.

**Comment by perryjrandall** [comment]
I merged this forcibly because the smoke tests were broken

```
        PASS [  65.698s] smoke-test aptos_cli::validator::test_join_and_leave_validator
------------
     Summary [  65.699s] 1 tests run: 1 passed (1 slow), 64 skipped
perry@Perrys-MBP aptos-node % LOCAL_SWARM_NODE_RELEASE=1 cargo nextest run --release --profile ci --package smoke-test --test-threads 6 --retries 3 -- aptos_cli::validator::test_join_and_leave_validator
```

File: network/netcore/src/transport/tcp.rs [hunk 14]
```diff
@@ -69,9 +106,23 @@ impl Transport for TcpTransport {
             return Err(invalid_addr_error(&addr));
         }
 
-        let listener = ::std::net::TcpListener::bind((ipaddr, port))?;
-        listener.set_nonblocking(true)?;
-        let listener = TcpListener::try_from(listener)?;
+        let addr = SocketAddr::new(ipaddr, port);
+
```
Comment (bchocho) [inline network/netcore/src/transport/tcp.rs:110]: Did we change this to not set_nonblocking?

**Pull request merged**